### PR TITLE
history-view: simplify code and add clear option

### PR DESCRIPTION
**Pull request by mbkma**
Inspired by `gnome-calculators` implementation. If this PR gets merged, it would make #139 obsolete. If you want, I can change the "Clear History" shortcut and menu item, I took Shift+Ctrl+C for now.
Screenshot at 2020-03-23 00-28-50
EDIT: Just noticed that Shift+Ctrl+C may be a bad idea, because some are used to copy like this from a terminal window. Maybe Ctrl+ESC would be more appropriate?

### CODE REVIEW

**Comment by sc0w** [comment]
> EDIT: Just noticed that Shift+Ctrl+C may be a bad idea, because some are used to copy like this from a terminal window. Maybe Ctrl+ESC would be more appropriate?

maybe <kbd>ctrl</kbd><kbd>shift</kbd><kbd>supr</kbd> is better like firefox and chromium does?

File: src/math-window.c [hunk 1]
```diff
@@ -509,6 +514,8 @@ static void create_menu(MathWindow* window)
     gtk_widget_add_accelerator(menu_item, "activate", accel_group, GDK_KEY_Z, GDK_CONTROL_MASK, GTK_ACCEL_VISIBLE);
     menu_item = add_menu_item(menu, gtk_image_menu_item_new_from_icon("edit-redo",_("_Redo"), accel_group), G_CALLBACK(redo_cb), window);
     gtk_widget_add_accelerator(menu_item, "activate", accel_group, GDK_KEY_Z, GDK_CONTROL_MASK | GDK_SHIFT_MASK, GTK_ACCEL_VISIBLE);
+    menu_item = add_menu_item(menu, gtk_image_menu_item_new_from_icon("gtk-clear",_("_Clear History"), accel_group), G_CALLBACK(clear_history_cb), window);
```
Comment (sc0w) [inline src/math-window.c]: `gtk-clear` is deprecated, you can use `edit-clear` instead

**Comment by sc0w** [comment]
New warnings in the logs:

```
math-history-entry.c:44:31: warning: unused parameter 'widget' [-Wunused-parameter]
   44 | answer_clicked_cb (GtkWidget *widget, GdkEventButton *eventbutton, gpointer data)
      |                    ~~~~~~~~~~~^~~~~~

math-history-entry.c:56:33: warning: unused parameter 'widget' [-Wunused-parameter]
   56 | equation_clicked_cb (GtkWidget *widget, GdkEventButton *eventbutton, gpointer data)
      |                      ~~~~~~~~~~~^~~~~~

math-window.c:183:43: warning: unused parameter 'menu' [-Wunused-parameter]
  183 | static void clear_history_cb(GtkMenuItem *menu, MathWindow *window)
      |                              ~~~~~~~~~~~~~^~~~
```

**Comment by mbkma** [comment]
@sc0w how did you generate those? I see them in travis, but I don't know the command to generate those warnings locally.

**Comment by sc0w** [comment]
@mbkma I can see all the `"unused parameter"` warnings here locally.

Anyway you can build explicity adding the cflag `-Wunused-parameter`

btw, you can see the cflags using by travis here:
https://github.com/mate-desktop/mate-calc/blob/master/.build.yml#L76

**Comment by raveit65** [comment]
@mbkma 
You need to update to mate-common-1.23.3 to see new warning levels. See git log
https://github.com/mate-desktop/mate-common/commits/master
Or better mate-common-1.24.0.
New default for mate-calc is `MATE_COMPILE_WARNINGS(maximum)`
https://github.com/mate-desktop/mate-calc/blob/master/configure.ac#L16

```
  --enable-compile-warnings=[no/minimum/yes/maximum/error]
                          Turn on compiler warnings
```

Personal i am not sure that `maximum` is the right choice as default for local compiling, but that's another discussion. See https://github.com/mate-desktop/mate-panel/pull/1050
Sadly, we don't have an agreement to handle that problem in general :/

**Comment by mbkma** [comment]
Ok, got it. For now I build with `./autogen.sh CFLAGS='-Wall -Wextra'`.
Concerning the new warnings:
First option: I could add G_GNUC_UNUSED, but this would be needed by a lot callback functions in mate-calc (see the amount of "unused parameter" warnings).
Second option: I leave it as it is.
I don't see another option because the signature of these callback functions is determined.

**Comment by raveit65** [comment]
Isn't there a third option to fix the warning instead to silent them?
I don't get why we first use -Wextra with `-Wunused` as default, and in a second step we silent them in code of every package.
If we are not able to fix the warnings than we should silent them in general for mate.
See https://github.com/mate-desktop/mate-panel/pull/1050#issuecomment-576718035
> This looks less readable to me... and the warning is minor, why did they include it in -Wextra? confused
> Would adding -Wno-unused-parameter after -Wextra work? The manual says that each warning option has such "no" switch to turn it off.

My 2cents

**Comment by sc0w** [comment]
@mbkma

> First option: I could add G_GNUC_UNUSED, but this would be needed by a lot callback functions in mate-calc (see the amount of "unused parameter" warnings).
> Second option: I leave it as it is.

Third option with (void):

https://github.com/mate-desktop/engrampa/pull/373

or maybe with `'#define UNUSED(x) (void)(x)'` an later with `'UNUSED (varname);'`

anyway, I think we can ignore all the "unused parameter" warnings in all the PRs by the moment, because as @raveit65 said:

> Sadly, we don't have an agreement to handle that problem in general :/

File: src/math-history.c [hunk 2]
```diff
@@ -98,20 +119,23 @@ static void
...
     history->priv->listbox = gtk_list_box_new();    
+
     gtk_list_box_set_selection_mode(GTK_LIST_BOX(history->priv->listbox), GTK_SELECTION_NONE);
     gtk_widget_show(GTK_WIDGET(history->priv->listbox));
 
     gtk_container_add(GTK_CONTAINER(history), history->priv->listbox);
     gtk_widget_set_valign(GTK_WIDGET(history->priv->listbox), GTK_ALIGN_END);
     gtk_widget_set_size_request(GTK_WIDGET(history), 100, 100);
-    gtk_widget_set_can_focus(GTK_WIDGET(history), FALSE);
+    gtk_widget_set_can_focus(GTK_WIDGET(history), false);
```
Comment (sc0w) [inline src/math-history.c]: why this change?

File: src/math-history.c [hunk 3]
```diff
@@ -50,33 +50,54 @@ scroll_bottom_cb(MathHistory *history, gpointer data)
...
-        }
+        g_free(answer_four_digits);
+        g_free(answer_nine_digits);
+        return;
     }
+
+    math_history_entry_insert_entry(entry, equation, answer_four_digits, answer_nine_digits);
+
+    gtk_list_box_insert(GTK_LIST_BOX(history->priv->listbox), GTK_WIDGET(entry), -1);
+    gtk_widget_set_can_focus(GTK_WIDGET(entry), false);
```
Comment (sc0w) [inline src/math-history.c]: FALSE instead false to make it more readable, please

Comment (mbkma) [reply src/math-history.c]: you are right

**Comment by sc0w** [comment]
@mbkma LGTM, please squash all the commits into one

**Comment by raveit65** [comment]
I just opened a PR to solve the problems with `-Wunused-parameter` warnings.
https://github.com/mate-desktop/mate-common/pull/31

**Comment by rbuj** [comment]
Bind sensitive of `Calculator -> Clear History` with active of `View -> History`.
```patch
diff --git a/src/math-window.c b/src/math-window.c
index ba55692..957a399 100644
--- a/src/math-window.c
+++ b/src/math-window.c
@@ -479,6 +479,7 @@ static void create_menu(MathWindow* window)
     GtkAccelGroup* accel_group;
     GtkWidget* menu;
     GtkWidget* menu_item;
+    GtkWidget* menu_item_clear_history;
     GSList* group = NULL;
 
     accel_group = gtk_accel_group_new();
@@ -514,8 +515,8 @@ static void create_menu(MathWindow* window)
     gtk_widget_add_accelerator(menu_item, "activate", accel_group, GDK_KEY_Z, GDK_CONTROL_MASK, GTK_ACCEL_VISIBLE);
     menu_item = add_menu_item(menu, gtk_image_menu_item_new_from_icon("edit-redo",_("_Redo"), accel_group), G_CALLBACK(redo_cb), window);
     gtk_widget_add_accelerator(menu_item, "activate", accel_group, GDK_KEY_Z, GDK_CONTROL_MASK | GDK_SHIFT_MASK, GTK_ACCEL_VISIBLE);
-    menu_item = add_menu_item(menu, gtk_image_menu_item_new_from_icon("edit-clear",_("_Clear History"), accel_group), G_CALLBACK(clear_history_cb), window);
-    gtk_widget_add_accelerator(menu_item, "activate", accel_group, GDK_KEY_Delete, GDK_CONTROL_MASK | GDK_SHIFT_MASK, GTK_ACCEL_VISIBLE);
+    menu_item_clear_history = add_menu_item(menu, gtk_image_menu_item_new_from_icon("edit-clear",_("_Clear History"), accel_group), G_CALLBACK(clear_history_cb), window);
+    gtk_widget_add_accelerator(menu_item_clear_history, "activate", accel_group, GDK_KEY_Delete, GDK_CONTROL_MASK | GDK_SHIFT_MASK, GTK_ACCEL_VISIBLE);
     add_menu_item(menu, gtk_separator_menu_item_new(), NULL, NULL);
     add_menu_item(menu, gtk_image_menu_item_new_from_icon("preferences-desktop",_("_Preferences"), accel_group), G_CALLBACK(show_preferences_cb), window);
     add_menu_item(menu, gtk_separator_menu_item_new(), NULL, NULL);
@@ -536,6 +537,7 @@ static void create_menu(MathWindow* window)
     menu = add_menu(window->priv->menu_bar, VIEW_MENU_LABEL);
     window->priv->view_history_menu_item = add_menu_item(menu, gtk_check_menu_item_new_with_mnemonic(VIEW_HISTORY_LABEL), G_CALLBACK(history_check_toggled_cb), window);
     gtk_widget_add_accelerator(window->priv->view_history_menu_item, "activate", accel_group, GDK_KEY_H, GDK_CONTROL_MASK, GTK_ACCEL_VISIBLE);
+    g_object_bind_property (window->priv->view_history_menu_item, "active", menu_item_clear_history, "sensitive", G_BINDING_DEFAULT | G_BINDING_SYNC_CREATE);
 
     menu = add_menu(window->priv->menu_bar, HELP_MENU_LABEL);
     menu_item = add_menu_item(menu, gtk_image_menu_item_new_from_icon("help-browser", HELP_CONTENTS_LABEL, accel_group), G_CALLBACK(help_cb), window);
```

**Comment by rbuj** [comment]
You would move `Clear History `menu item into `View` menu, since the history stack differs from the undo/redo stacks.

**Comment by mbkma** [comment]
> You would move `Clear History `menu item into `View` menu, since the history stack differs from the undo/redo stacks.

That is what I did first, but I don't know how to solve this, which seems to happen in `gtk_menu_shell_append`:

Screenshot at 2020-04-06 16-57-07

**Comment by sc0w** [comment]
clear history looks fine in the "calculator" menu, I don't understand why must to move to "view"

**Comment by sc0w** [comment]
> Bind sensitive of Calculator -> Clear History with active of View -> History.

With the change you suggested, it will only work if history is on?

why?

It should always work in my opinion.
